qutip/qip/qasm.py

```diff
@@ -0,0 +1,400 @@
+                with open(filename, "r") as f:
+                    qasm_lines = [line.strip() for line in f.read().splitlines()]
+                    qasm_lines = list(filter(lambda x: x[:2] != "//" and x != "",
+                                    qasm_lines))
+
+                    expanded_commands = (expanded_commands
+                                        + command[prev_index:curr_index]
+                                        + _tokenize(qasm_lines))
+                    prev_index = curr_index + 1
+            else:
+                raise ValueError(command[1] + ": such a file does not exist")
+
+        expanded_commands += self.commands[prev_index:]
+        self.commands = expanded_commands
+
+    def _custom_gate(self, qc_temp, gate_call):
+        '''
+        Recursively processes a custom-defined gate with specified arguments
+        to produce a dummy circuit with all the gates constituting it.
+        '''
+
+        gate_name, args, regs = gate_call
+        gate = self.qasm_gates[gate_name]
+        args_map = {}
+        regs_map = {}
+
+        # maps variables to supplied arguments, registers
+        for i, arg in enumerate(gate.gate_args):
+            args_map[arg] = eval(str(args[i]))
+        for i, reg in enumerate(gate.gate_regs):
+            regs_map[reg] = regs[i]
+
+        # process all the constituent gates with supplied arguments, registers
+        for call in gate.gates_inside:
+
+            # create function call for the constituent gate
+            name, com_args, com_regs = call
+            for arg, real_arg in args_map.items():
+                com_args = [command.replace(arg.strip(),
+                                            str(real_arg)) for command
+                                                            in com_args]
+            for reg, real_reg in regs_map.items():
+                com_regs = [command.replace(reg.strip(),
+                                            str(real_reg)) for command
+                                                            in com_regs]
+            com_args = [eval(arg) for arg in com_args]
+
+            if name == "CX":
+                qc_temp.add_gate("CNOT",  targets=int(com_regs[1]),
+                            controls=int(com_regs[0]))
+            elif name == "U":
+                target = int(com_regs[0])
+                qc_temp.add_gate("RZ", targets=target,
+                            arg_value=float(com_args[2]))
+                qc_temp.add_gate("RY", targets=target,
+                            arg_value=float(com_args[0]))
+                qc_temp.add_gate("RZ", targets=target,
+                            arg_value=float(com_args[1]))
+            else:
+                self._custom_gate(qc_temp, [name, com_args, com_regs])
+
+    def _regs_processor(self, regs, reg_type):
+        '''
+        Processes register tokens mapping them to the QubitCircuit index
+        of the respective registers.
```

Comment:
Comments should be read like a command to the code that follows. I don't where I read this, but I am sure it's in [pep8](https://www.python.org/dev/peps/pep-0008/#comments) somewhere.
(I want to do that thing that @nathanshammah showed us where we make suggested changes. Cant see the option)
`Process the register tokens: map them to the QubitCircuit index of the respective registers.`